design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,410 @@
+To add the support for the brand new storage backend `foostore`, it should be added to the `alpha` CRD because that properly represents its maturity.
+Therefore, we would have the following yaml:
+
+```yaml
+apiVersion: rook.io/v1
+kind: StorageCluster
+spec:
+  someStableField: "stableValue"
+  ceph:
+  mino:
+---
+apiVersion: rook.io/v1alpha1
+kind: StorageCluster
+spec:
+  someAlphaField: "alphaValue"
+  ceph:
+  minio:
+  foostore:
+```
+
+This results in an unfortunate situation for the new `foostore` backend.
+Because its version is coupled with that of the top level cluster CRD, and it is of `alpha` maturity, it is forced to be included in the old `alpha` cluster type, meaning it can't use the latest mature and stable fields.
+This feels like a very negative restriction and suggests that we should decouple storage backends from common cluster abstractions.
+
+Commonality and sharing of types and implementations is important and is another way Rook provides value to the greater storage community beyond the operators that it implements.
+To facilitate this, the following API groups are proposed:
+
+* `rook.io`: common abstractions that have use across multiple storage backends and types
+* `storage.rook.io`: specific storage backend types, such as Ceph.
+
+This would give us the following example manifests that a user might deploy, note how each backend has its own independent version:
+
+```yaml
+apiVersion: storage.rook.io/v1
+kind: Ceph
+spec:
+  cluster: # common cluster scope, properties, etc.
+  monCount: 3
+---
+apiVersion: storage.rook.io/v1beta1
+kind: Minio
+spec:
+  cluster: # common cluster scope, properties, etc.
+  mode: distributed
+---
+apiVersion: storage.rook.io/v1apha2
+kind: Foostore
+spec:
+  cluster: # common cluster scope, properties, etc.
+  fooField: "fooValue"
+```
+
+The `cluster` field used in the preceding example can be for **any** version of the common cluster abstraction that the storage backend wants to reference.
+For example, an `alpha` Foostore type could reference a stable cluster type like so in `types.go`:
+
+```go
+package v1alpha1
+
+import (
+  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+  v1 "github.com/rook/rook/pkg/apis/rook.io/v1"
+)
+
+type Foostore struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata"`
+  Spec              FoostoreSpec   `json:"spec"`
+  Status            FoostoreStatus `json:"status,omitempty"`
+}
+
+type FoostoreSpec struct {
+  ClusterSpec v1.ClusterSpec `json:"cluster"`
+  FooField    string         `json:"fooField"`
+}
+```
+
+#### StorageCluster
+
+The `StorageCluster` type provides the general abstraction for a distributed storage cluster.
+The spec for this type will include image version, networking information, placement information, resource usage requests, etc.
+It will also contain a `StorageScope` type so that the boundaries of the cluster's storage can be defined.
+
```

Comment:
could we combine storage cluster and storage scope?

Reply:
We may be able to, but I was thinking of keeping `StorageScope` as a separate type so that it could be reused easily in other cloud-native storage types that aren't distributed storage clusters.